src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,159 @@
+        } else {
+            Node<T> nodeAfterCurrent = first;
+            nodeAfterCurrent.next.prev = nodeAfterCurrent;
+            nodeAfterCurrent = findNode(nodeAfterCurrent, index);
+            Node<T> nodeBefore = nodeAfterCurrent.prev;
+            Node<T> current = new Node<>(nodeAfterCurrent.prev, value, nodeAfterCurrent);
+            nodeAfterCurrent.prev = current;
+            nodeBefore.next = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = first;
+        currentNode.next.prev = currentNode;
+        currentNode = findNode(currentNode, index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeSet = first;
+        nodeSet.next.prev = nodeSet;
+        nodeSet = findNode(nodeSet, index);
+        T oldValue = nodeSet.value;
+        nodeSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T result;
+        if (index == 0) {
+            result = first.value;
+            first = first.next;
+            size--;
+        } else if (index == size - 1) {
+            result = last.value;
+            last.prev.next = null;
+            last = last.prev;
+            size--;
+        } else {
+            Node<T> nodeDeleted = first;
+            nodeDeleted.next.prev = nodeDeleted;
+            nodeDeleted = findNode(nodeDeleted, index);
+            result = nodeDeleted.value;
+            nodeDeleted.prev.next = nodeDeleted.next;
+            nodeDeleted.next.prev = nodeDeleted.prev;
+            size--;
+        }
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int count = 0;
+        Node<T> nodeRemoved = first;
+        while (count != size && nodeRemoved.value != object) {
+            if (nodeRemoved.value != null && nodeRemoved.value.equals(object)) {
+                count++;
+                remove(count);
+                return true;
+            }
+            nodeRemoved = nodeRemoved.next;
+            count++;
+        }
+        if (count == size) {
+            return false;
+        }
+        remove(count);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return first == null;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private Node<T> findNode(Node<T> node, int index) {
```

Comment:
[checklist](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html): Don’t iterate your nodes from the head to tail or vice versa.

Reply:
still in progress - as I feel - for this I need to work with this
"
  Node<T> nodeSet = first;
  nodeSet.next.prev = nodeSet;
"

Reply:
for this momnet I can not understand how to start surfing in LinkedList fm middle Node...but I am thinking